### ISSUE TITLE
Mobile-ffmpeg does not support multi instance of FFmpeg

### ISSUE THREAD

**Issue by at-kietvo** [tanersener/mobile-ffmpeg#118]
Hello @tanersener, I have been choosing your library for fully integrate in my project, all of this working well. But I have the issue when using multiple thread for convert multiple files as the same time. FFmpeg class can't handle this case, when I call `ffmpeg.execute` in multi-thread, its crash on my device, and never see the log/error message.

Checking for the deep of library, I think the `FFmpeg` just support one instance as on, I try to limit the pool of multi-thread to `1` and this work.

So, do you have any idea/suggestion in this case, please let me know. 
Thanks.

**Environment**
 - Platform: Android
 - Architecture: [arm-v7a, arm-v7a-neon, arm64-v8a, x86, x86_64, armv7, armv7s, arm64, i386, x86_64]

**Comment by tanersener** [comment]
Current implementation do support threads but does not support parallel executions. Only one execution is possible at any particular time.

[Creating Multiple Outputs](https://trac.ffmpeg.org/wiki/Creating%20multiple%20outputs) wiki page includes different options to encode multiple files at once. You can try suggestions given on that page and see if one of them can help you overcome this limitation.

**Comment by at-kietvo** [comment]
Yes, this link seem to work, but It can't handle the event when the each output is complete, something like when i have 10 input files and return 10 output files, this process will be finished when all of output files returned. We can't deliver progress indicator in this case.

So, my stupid question: Is this library (.so file also JNI) can create multiple instance of FFmpeg or not. Becasue I don't have many time to take a look in the source code, if it can, in the shortly future, I will try to do my best.

Thanks for your quickly answer.

**Comment by tanersener** [comment]
> So, my stupid question: Is this library (.so file also JNI) can create multiple instance of FFmpeg or not.

Unfortunately no, it can't create multiple `FFmpeg` instances.

**Comment by function1983** [comment]
Hi @tanersener,

Thank you for your project.
I have my own working implementation of multiple ffmpeg instances, all pointing to the same ffmpeg.so, running stable on multiple android device types (TVBox, phones) 

The setup is as followed:
```(4x) rtsp streams -> (4x) ffmpeg instances (running on the same android device) -> (4x) files save to disk.```

Do you think this is @at-kietvo's initial question? Because at first I think his setup is:
**Scenario A**: ```(10x) files -> (10x) ffmpeg instances -> (10x) files save to disk```
But after reading the discussion, now I think his requirement is:
**Scenario B**: ```(10x) files -> (1x) ffmpeg instance -> (10x) files save to disk```

If he wants to accomplish **Scenario A**, then technically it should be possible, don't you think?

**Comment by tanersener** [comment]
Hi @function1983, which binary files and shared libraries do you have in your application bundle? How do you execute a command using them?

**Comment by function1983** [comment]
Hi @tanersener , sure thing.
Mine is a modified version of 'nl.bravobit:android-ffmpeg:1.1.5'. 
[https://github.com/bravobit/FFmpeg-Android/blob/master/android-ffmpeg/src/main/java/nl/bravobit/ffmpeg/FFcommandExecuteAsyncTask.java](url)

```
ExecutorService myExecutor = Executors.newFixedThreadPool(4);
FFcommandExecuteAsyncTask task = new FFcommandExecuteAsyncTask(command, timeout, ffmpegExecuteResponseHandler);
            task.executeOnExecutor(myExecutor);
```
Quite a few modern chips with ffmpeg HEVC hwaccel support just hit the market recently. Bravobit's library doesn't support HEVC but looks like yours does so I am really keen on trying it out.

**Comment by tanersener** [comment]
Thanks @function1983, 

As you can see from [this folder](https://github.com/bravobit/FFmpeg-Android/tree/master/android-ffmpeg/src/main/assets/x86), `nl.bravobit:android-ffmpeg` uses `ffmpeg` binary. When you run a command you actually start another process which executes `ffmpeg` binary. Each process runs on a different memory address and you can start as many processes as your device allows to.

Unfortunately, in `mobile-ffmpeg` I don't generate any executables and I only have shared libraries, .so files. When you run a command, functions defined in shared libraries are called by the same process of your application, in the same memory space of your application. When multiple commands are started then all library functions are executed in the same memory context. Shared libraries don't support that and crash at one point. Technically, it is not possible to initiate a new process using shared libraries in Android. Because of that I can't do anything to support parallel executions.

Sadly, if you try to run four commands using my library your application will crash.

**Comment by function1983** [comment]
Wow, thanks for the very detailed explanation. I thought both use the .so library route.
Still, your project looks very promising. I will take an in depth look at it when I have more time. Keep up the good work!

**Comment by codaman** [comment]
i solved this to create a queue. ffmpeg uses large resources. If you will run multiple instance, your device couldn't  handle this. I think you can choose queue option

**Comment by herksaw** [comment]
Personally I think it is still helpful to have multiple instances of ffmpeg available. In my case, I would need to watch the RMS changes constantly in a live stream, while cropping part of the stream with certain length once it reached the RMS threshold. The second ffmpeg execution will stop the first one consequently, it is perfectly fine to connect to the stream again, but I'm thinking for a better way.

**Comment by tanersener** [comment]
This feature, `Concurrent execution support`, is implemented in `v4.3.1`. You can use it in the latest versions.

See the CHANGELOG of [v4.3.1](https://github.com/tanersener/mobile-ffmpeg/wiki/Changelog#v431---25-01-2020).
